// @tago-io/sdk/out/modules/Account/Actions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class Actions extends TagoIOModule_1.default {
    /**
     * Retrieves a list with all actions from the account
     * @default
     * ```json
     * queryObj: {
     *   page: 1,
     *   fields: ["id", "name"],
     *   filter: {},
     *   amount: 20,
     *   orderBy: "name,asc",
     * }
     * ```
     * @param queryObj Search query params
     */
    list(queryObj) {
        const result = this.doRequest({
            path: "/action",
            method: "GET",
            params: {
                page: (queryObj === null || queryObj === void 0 ? void 0 : queryObj.page) || 1,
                fields: (queryObj === null || queryObj === void 0 ? void 0 : queryObj.fields) || ["id", "name"],
                filter: (queryObj === null || queryObj === void 0 ? void 0 : queryObj.filter) || {},
                amount: (queryObj === null || queryObj === void 0 ? void 0 : queryObj.amount) || 20,
                orderBy: (queryObj === null || queryObj === void 0 ? void 0 : queryObj.orderBy) ? `${queryObj.orderBy[0]},${queryObj.orderBy[1]}` : "name,asc",
            },
        });
        return result;
    }
    /**
     * Generates and retrieves a new action from the account
     * @param actionObj Object data to create new TagoIO Action
     */
    create(actionObj) {
        const result = this.doRequest({
            path: "/action",
            method: "POST",
            body: actionObj,
        });
        return result;
    }
    /**
     * Modify any property of the action.
     * @param actionID Action ID
     * @param actionObj Action Object with data to be replaced
     */
    edit(actionID, actionObj) {
        const result = this.doRequest({
            path: `/action/${actionID}`,
            method: "PUT",
            body: actionObj,
        });
        return result;
    }
    /**
     * Deletes an action from the account
     * @param actionID Action ID
     */
    delete(actionID) {
        const result = this.doRequest({
            path: `/action/${actionID}`,
            method: "DELETE",
        });
        return result;
    }
    /**
     * Gets information about the action
     * @param actionID Action ID
     */
    info(actionID) {
        const result = this.doRequest({
            path: `/action/${actionID}`,
            method: "GET",
        });
        return result;
    }
}
exports.default = Actions;
