// @tago-io/sdk/out/modules/Account/Plan.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class Plan extends TagoIOModule_1.default {
    /**
     * Active plan and set services limit
     * @param data
     * @example
     * data: {
     *   "plan": "plan_id",
     *   "sms": 100,
     *   "email": 1000,
     *   "data_records": 200000,
     *   "device_request": 250,
     *   "analysis": 1000
     * }
     */
    async setPlanParameters(data) {
        const result = await this.doRequest({
            path: "/account/plan/",
            method: "POST",
        });
        return result;
    }
    /**
     * Get Active Plan and Services
     */
    async getActivePlan() {
        const result = await this.doRequest({
            path: "/account/plan",
            method: "GET",
        });
        return result;
    }
    /**
     * Shows a summary of how much your account is costing, divided by sections
     */
    async summary() {
        const result = await this.doRequest({
            path: "/billing",
            method: "GET",
        });
        return result;
    }
    /**
     * Get current Tago pricing
     */
    async getCurrentPrices() {
        const result = await this.doRequest({
            path: "/pricing",
            method: "GET",
        });
        return result;
    }
}
exports.default = Plan;
